lib/pbio/drv/block_device/block_device_ev3.c

```diff
@@ -0,0 +1,730 @@
+        if (err != PBIO_SUCCESS) {
+            return err;
+        }
+        PBIO_OS_AWAIT_WHILE(state, bdev.spi_status & SPI_STATUS_WAIT_ANY);
+
+        // Erase this block
+        set_address_be(&erase_address[1], PBDRV_CONFIG_BLOCK_DEVICE_EV3_START_ADDRESS + offset);
+        err = spi_begin_for_flash(erase_address, sizeof(erase_address), 0, 0, 0);
+        if (err != PBIO_SUCCESS) {
+            return err;
+        }
+        PBIO_OS_AWAIT_WHILE(state, bdev.spi_status & SPI_STATUS_WAIT_ANY);
+
+        // Wait for completion
+        PBIO_OS_AWAIT(state, &sub, err = flash_wait_write(&sub));
+        if (err != PBIO_SUCCESS) {
+            return err;
+        }
+    }
+
+    // Write page by page.
+    for (size_done = 0; size_done < size; size_done += size_now) {
+        size_now = pbio_int_math_min(size - size_done, FLASH_SIZE_WRITE);
+
+        // Enable writing
+        err = spi_begin_for_flash(cmd_write_enable, sizeof(cmd_write_enable), 0, 0, 0);
+        if (err != PBIO_SUCCESS) {
+            return err;
+        }
+        PBIO_OS_AWAIT_WHILE(state, bdev.spi_status & SPI_STATUS_WAIT_ANY);
+
+        // Write this block
+        set_address_be(&write_address[1], PBDRV_CONFIG_BLOCK_DEVICE_EV3_START_ADDRESS + size_done);
+        err = spi_begin_for_flash(write_address, sizeof(write_address), buffer + size_done, 0, size_now);
+        if (err != PBIO_SUCCESS) {
+            return err;
+        }
+        PBIO_OS_AWAIT_WHILE(state, bdev.spi_status & SPI_STATUS_WAIT_ANY);
+
+        // Wait for completion
+        PBIO_OS_AWAIT(state, &sub, err = flash_wait_write(&sub));
+        if (err != PBIO_SUCCESS) {
+            return err;
+        }
+    }
+
+    PBIO_OS_ASYNC_END(PBIO_SUCCESS);
+}
+
+static pbio_os_process_t pbdrv_block_device_ev3_init_process;
+
+pbio_error_t pbdrv_block_device_ev3_init_process_thread(pbio_os_state_t *state, void *context) {
+    pbio_error_t err;
+
+    PBIO_OS_ASYNC_BEGIN(state);
+
+    // Write the ID getter command
+    err = spi_begin_for_flash(cmd_rdid, sizeof(cmd_rdid), 0, 0, 0);
+    if (err != PBIO_SUCCESS) {
+        return err;
+    }
+    PBIO_OS_AWAIT_WHILE(state, bdev.spi_status & SPI_STATUS_WAIT_ANY);
+
+    // Verify flash device ID
+    if (memcmp(device_id, &bdev.spi_cmd_buf_rx[1], sizeof(device_id))) {
+        return PBIO_ERROR_FAILED;
+    }
+
+    // Initialization done.
+    pbdrv_init_busy_down();
+
+    PBIO_OS_ASYNC_END(PBIO_SUCCESS);
+}
+
+void pbdrv_block_device_init(void) {
+    // SPI module basic init
+    PSCModuleControl(SOC_PSC_0_REGS, HW_PSC_SPI0, PSC_POWERDOMAIN_ALWAYS_ON, PSC_MDCTL_NEXT_ENABLE);
+    SPIReset(SOC_SPI_0_REGS);
+    SPIOutOfReset(SOC_SPI_0_REGS);
+    SPIModeConfigure(SOC_SPI_0_REGS, SPI_MASTER_MODE);
+    unsigned int  spipc0 = SPI_SPIPC0_SOMIFUN | SPI_SPIPC0_SIMOFUN | SPI_SPIPC0_CLKFUN | SPI_SPIPC0_SCS0FUN0 | SPI_SPIPC0_SCS0FUN3;
+    SPIPinControl(SOC_SPI_0_REGS, 0, 0, &spipc0);
+    SPIDefaultCSSet(SOC_SPI_0_REGS, (1 << 0) | (1 << 3));
```

Comment:
Let's use macros here so we know what these bits are.

`SPI_DATA_FORMATx` are not really better though. I would introduce a new one that says `SPI_BITS_PER_WORD(x)`